Assignment-3/README.md

```diff
@@ -0,0 +1,66 @@
+resiliency to write into my code. This submission reflects my best efforts. Reflecting 
+on what I have done, I think I built too much resiliency (too many exception checks)
+into my first two programs (1.1, 1.2). I hope that the later programs still have
+enough, but I would appreciate any feedback on this points.
+
+The following section gives an overview of certain codes in this Assignment.
+This documentation is not holistic, and should be considered in conjunction with
+comments provided in-line throughout the code.
+
+Documentation for the codes in the Assignment-3 Folder:
+
+1.1_db_streamT.py
+Mostly the same as Assignment-2, this code stores the tweet data into MongoDB. 
+This code was not re-run while completing Assignment-3 as the MongoDB already
+existed from Assignment-2.
+
+1.2_db_tweets.py
+This code loads tweets from the S3 JSON files stored in Assignment-2, then parses
+out the text, saving it into a MongoDB named db_tweets. I have decided to store
+unicode as received from Twitter so that this code is flexible to tweets in a variety 
+of languages.
+
+2.1_top_30.py
+This code uses the MongoDB aggregate function to find the top 30 tweets based on the
+tweet content. It then pulls data about the original poster of the tweet from db_streamT
+
+2.2_lexical_div.py
+This code computes the lexical diversity for each user for whom we collected tweets. In this analysis,
+we allocate retweet content to the user who retweeted it. There are many assumptions in this analysis,
+spelled out at the start of the code. See "lexical_diversity_hist.txt" for the requested plot
+of the results. We find that many users had a diversity of 1 (only used unique words). We hypothesize
+that these are individuals who only tweeted about #Microsoft or #Mojang once in the study period, thus
+they did not have time to use multiple instances of words. After this, we see a "hump" in the lower 90%,
+and 80%s. Again, with the small number of tweets per user we should be suspicious of drawing too much
+from this analysis without further investigation.
+
+2.3_db_followers.py
+This code takes the top 30 tweets stored during the run of program 2.1, and pulls the followers.
+It then waits a week, and pulls the followers again. Finally, the program compares the followers
+and identify "unfollowers". We complete most of this task in storage as opposed to memory by storing 
+to MongoDB, and using the "aggregate" function (and its pipeline) to identify users who were followers 
+at time 1 but not at time 2
+
+3.1_S3_store.py
+Stores db_streamT and db_tweets to S3 in bucket com.christopherllop.w205assignment3
+Note: I spent many hours trying to get the mongodump / mongorestore functions to
+work, but was unable to figure out how to do so. I would greatly appreciate some 
+feedback regarding how to implement this given my setup. The task seems to require
+functions outside of the scope of the class - but I would still like to learn them.
```

Comment:
Though you can use mongodump/mongostore commands, there are other ways to implement this: Here is an example code that uploads the backup of db_tweets to S3

my_db = conn['db_tweets']
my_con = boto.connect_s3(aws_key, aws_secret)
my_buck = my_con.get_bucket('W205')
S3_out = Key(my_buck)

S3_out.key = 'db_tweets.json'
S3_out.set_contents_from_string(dumps(my_db.tweets.find()))

2- Using the mongodump to create a backup:
You can use the subprocess class:

import shlex
dbname='db_tweets'
db_backup_path='./back/'
 command_to_execute = "mongodump --db "+dbname+" --out "+db_backup_path
process_command = shlex.split( command_to_execute)
process = subprocess.Popen(process_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
stdout, stderr = process.communicate()

More information about the subprocess class can be found:
https://docs.python.org/2/library/subprocess.html